tests/ops/functions/test_matrix.py

```diff
@@ -0,0 +1,580 @@
+        assert np.allclose(res, expected)
+
+    @pytest.mark.parametrize("op_class", one_qubit_one_parameter)
+    def test_parametric_gates_instantiated(self, op_class):
+        """Verify that the matrices of non-parametric one qubit gates is correct
+        when provided as an instantiated operation"""
+        op = op_class(0.54, wires=0)
+        res = qml.matrix(op)
+        expected = op.get_matrix()
+        assert np.allclose(res, expected)
+
+    @pytest.mark.parametrize("op_class", one_qubit_one_parameter)
+    def test_parametric_gates_qfunc(self, op_class):
+        """Verify that the matrices of non-parametric one qubit gates is correct
+        when provided as a qfunc"""
+        res = qml.matrix(op_class)(0.54, wires=0)
+        expected = op_class(0.54, wires=0).get_matrix()
+        assert np.allclose(res, expected)
+
+    @pytest.mark.parametrize("op_class", one_qubit_one_parameter)
+    def test_parametric_gates_qnode(self, op_class):
+        """Verify that the matrices of non-parametric one qubit gates is correct
+        when provided as a QNode"""
+        dev = qml.device("default.qubit", wires=1)
+        qnode = qml.QNode(lambda x: op_class(x, wires=0) and qml.probs(wires=0), dev)
+        res = qml.matrix(qnode)(0.54)
+        expected = op_class(0.54, wires=0).get_matrix()
+        assert np.allclose(res, expected)
+
+    @pytest.mark.parametrize("op_class", one_qubit_one_parameter)
+    def test_adjoint(self, op_class):
+        """Test that the adjoint is correctly taken into account"""
+        res = qml.matrix(qml.adjoint(op_class))(0.54, wires=0)
+        expected = op_class(-0.54, wires=0).get_matrix()
+        assert np.allclose(res, expected)
+
+    def test_ctrl(self):
+        """Test that the ctrl is correctly taken into account"""
+        res = qml.matrix(qml.ctrl(qml.PauliX, 0))(wires=1)
+        expected = CNOT
+        assert np.allclose(res, expected)
+
+    @pytest.mark.parametrize("target_wire", [0, 2, 3, 4])
+    def test_CNOT_permutations(self, target_wire):
+        """Test CNOT: 2-qubit gate with different target wires, some non-adjacent."""
+        res = qml.matrix(qml.CNOT, wire_order=[0, 1, 2, 3, 4])(wires=[1, target_wire])
+
+        # compute the expected matrix
+        perm = np.swapaxes(
+            np.swapaxes(np.arange(2**5).reshape([2] * 5), 0, 1), 0, target_wire
+        ).flatten()
+        expected = reduce(np.kron, [CNOT, I, I, I])[:, perm][perm]
+        assert np.allclose(res, expected)
+
+    def test_hamiltonian(self):
+        """Test that the matrix of a Hamiltonian is correctly returned"""
+        H = qml.PauliZ(0) @ qml.PauliY(1) - 0.5 * qml.PauliX(1)
+        mat = qml.matrix(H, wire_order=[1, 0, 2])
+        expected = reduce(np.kron, [Y, Z, I]) - 0.5 * reduce(np.kron, [X, I, I])
+
+    @pytest.mark.xfail(
+        reason="This test will fail because Hamiltonians are not queued to tapes yet!"
+    )

```

Comment:
Interesting!

Reply:
yes, I didn't realize this was the case either! something we should really fix.